asserts/messages.go

```diff
@@ -0,0 +1,170 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2025 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package asserts
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+)
+
+var (
+	validMessageID   = regexp.MustCompile(`^([a-zA-Z0-9]{4,16})(?:-(\d+))?$`)
+	validMessageKind = regexp.MustCompile(`^[a-z]+(?:-[a-z]+)*$`)
+)
+
+// DeviceID represents a unique device identifier composed of <brand-id>.<model>.<serial>.
+type DeviceID struct {
+	BrandID string
+	Model   string
+	Serial  string
+}
+
+// RequestMessage represents a request message assertion used to trigger actions on snapd.
+type RequestMessage struct {
+	assertionBase
+
+	id     string
+	seqNum *int
+
+	devices []DeviceID
+
+	assumes []string
+
+	sinceUntil
+	timestamp time.Time
+}
+
+// AccountID returns the account identifier that sent this request message.
+func (req *RequestMessage) AccountID() string {
+	return req.HeaderString("account-id")
+}
+
+// ID returns the message identifier without any sequence number suffix.
+func (req *RequestMessage) ID() string {
+	return req.id
+}
+
+// SeqNum returns the message's sequence number within a sequence, or nil if not sequenced.
+func (req *RequestMessage) SeqNum() *int {
+	return req.seqNum
+}
+
+// Kind returns the message kind that determines which subsystem handles the message.
+func (req *RequestMessage) Kind() string {
+	return req.HeaderString("message-kind")
+}
+
+// Devices returns the list of device IDs that this message targets.
+func (req *RequestMessage) Devices() []DeviceID {
+	return req.devices
+}
+
+// Assumes returns the list of device properties required for this message to be applied.
+func (req *RequestMessage) Assumes() []string {
+	return req.assumes
+}
+
+func assembleRequestMessage(assert assertionBase) (Assertion, error) {
+	accountID := assert.HeaderString("account-id")
+	if !validAccountID.MatchString(accountID) {
+		return nil, fmt.Errorf("invalid account id: %s", accountID)
+	}
+
+	rawID, err := checkStringMatches(assert.headers, "message-id", validMessageID)
+	if err != nil {
+		return nil, err
+	}
+
+	id := rawID
+	var seqNum *int
+	dashIdx := strings.LastIndex(rawID, "-")
+	if dashIdx != -1 {
+		id = rawID[:dashIdx]
+		seq, _ := strconv.Atoi(rawID[dashIdx+1:])
+		seqNum = &seq
+	}
```

Comment:
Usually we have sequence numbers start at 1, so `foo-0` is not allowed. We should add a check for that. 
That also helps in that we can use zero as the value returned when the sequence number is unset

Reply:
I checked the spec and it explicitly states that sequence numbers start at 1